### ISSUE TITLE
Cannot find module 'async'

### ISSUE THREAD

**Issue by guilhermevini** [binoculars/aws-lambda-ffmpeg#14]
I upload a file to S3 and lamda run, but with error.  

`{
  "errorMessage": "Cannot find module 'async'",
  "errorType": "Error",
  "stackTrace": [
    "Function.Module._resolveFilename (module.js:338:15)",
    "Function.Module._load (module.js:280:25)",
    "Module.require (module.js:364:17)",
    "require (module.js:380:17)",
    "Object.<anonymous> (/var/task/index.js:12:13)",
    "Module._compile (module.js:456:26)",
    "Object.Module._extensions..js (module.js:474:10)",
    "Module.load (module.js:356:32)",
    "Function.Module._load (module.js:312:12)",
    "Module.require (module.js:364:17)"
  ]
}`

**Comment by binoculars** [comment]
Do you have async in your `dist/node_modules/`? Run `zipinfo dist.zip` and see if it looks like this:

```
➜  aws-lambda-ffmpeg git:(master) zipinfo dist.zip
Archive:  dist.zip   33482786 bytes   15 files
-rw-r--r--  3.0 unx      402 b- defN 15-Feb-16 01:00 config.json
-rwxr-xr-x  3.0 unx 45920904 b- defN 15-Feb-16 01:00 ffmpeg
-rwxr-xr-x  3.0 unx 45843720 b- defN 15-Feb-16 01:00 ffprobe
-rw-r--r--  3.0 unx     5423 b- defN 15-Feb-16 01:00 index.js
drwxr-xr-x  3.0 unx        0 b- stor 15-Feb-16 01:00 node_modules/
drwxr-xr-x  3.0 unx        0 b- stor 15-Feb-16 01:00 node_modules/async/
-rw-r--r--  3.0 unx     5398 b- defN  7-Jan-16 19:03 node_modules/async/CHANGELOG.md
-rw-r--r--  3.0 unx     1063 b- defN 19-May-15 17:03 node_modules/async/LICENSE
-rw-r--r--  3.0 unx    60782 b- defN  7-Jan-16 18:18 node_modules/async/README.md
drwxr-xr-x  3.0 unx        0 b- stor 15-Feb-16 01:00 node_modules/async/dist/
drwxr-xr-x  3.0 unx        0 b- stor  7-Jan-16 17:59 node_modules/async/lib/
-rw-r--r--  3.0 unx    65171 b- defN 15-Feb-16 01:00 node_modules/async/package.json
-rw-r--r--  3.0 unx    38564 b- defN  7-Jan-16 19:04 node_modules/async/dist/async.js
-rw-r--r--  3.0 unx    12476 b- defN  7-Jan-16 19:04 node_modules/async/dist/async.min.js
-rw-r--r--  3.0 unx    38564 b- defN  7-Jan-16 17:59 node_modules/async/lib/async.js
15 files, 91992467 bytes uncompressed, 33480972 bytes compressed:  63.6%
```

**Comment by guilhermevini** [comment]
Yes! The package is normal. This error is strange. I checked many times. Maybe it's the version. You upgraded recently .

**Comment by binoculars** [comment]
I'm unable to reproduce this. I just clone the repo into a new directory and tried it from scratch. See this 
[gist](https://gist.github.com/binoculars/1b064a6681c19a4eef10) for details. I recommend starting over with a new clone of the repository. I also recommend using nvm with v0.10.36 while you're doing it. There may be an issue if you're not using the same version of Node.js as AWS Lambda is using.

```
➜  aws-lambda-ffmpeg git:(master) ✗ nvm current
v0.10.36
```

I need more information about your environment to resolve this. If you don't mind posting a gist similar to mine, that would help out a lot.

**Comment by binoculars** [comment]
Unable to reproduce this issue. Please re-open this if you can provide more details.

**Comment by akempe** [comment]
I've got the same issue, running under Windows. The issue appears to be related to the way gulp-zip is creating the zip file. If I manually compress the dist directory, then the async module is loaded fine. Looking at the directory structure in the two ZIPs, they appear identical - but clearly something isn't right in there. I don't know enough about gulp-zip to comment further - does this give you any ideas to the issue?

**Comment by akempe** [comment]
Looking more at this, it appears to be related to the way permissions bits are set on files in the archive. In Windows, there are none (which causes its own problems for the binaries as they're not marked as executable). I suspect the OP's issues were either Windows based or they had weird perms on the modules directory.

**Comment by binoculars** [comment]
Under the hood, gulp-zip uses yazl, which looks at the file's stats (See [this line in the code](https://github.com/thejoshwolfe/yazl/blob/master/index.js#L33)), and sets the permissions in the zipfile based on that. I would like to get this resolved, but don't want to reinvent the wheel to do it.

**Comment by akempe** [comment]
I want to get this sorted out too. Surely someone has bumped into this in the past?

Looking briefly at gulp-zip, I suspect the easiest way would be a change to allow an explicit mode / modes for specific files. Maybe check with them as to what their thoughts are - maybe they have come across this and have insight?

**Comment by akempe** [comment]
I've confirmed that explicitly setting file/dir perms when adding files to the archive (via changes to gulp-zip) fixes the issue. Now to find the most elegant way of doing this.

**Comment by akempe** [comment]
Reading more on gulp-zip, they have already said they don't think they should be setting these perms. I suggest that a change is made between gulp.src() and piping the files to zip() where the modes of all the files are explicitly set if they're missing.

**Comment by binoculars** [comment]
Ah good call. I was thinking we'd have to submit a PR for gulp-zip, but that makes it simpler. What about [sindresorhus/gulp-chmod](/sindresorhus/gulp-chmod)?

**Comment by akempe** [comment]
looks like it doesn't do directories (it explicitly skips them)

**Comment by akempe** [comment]
a pull request was submitted to fix this there https://github.com/sindresorhus/gulp-chmod/pull/4, as per gulp-chown, but there was no action. I agree with the maintainer that dir perms should be a separate option. If we can get that pull request redone and accepted, we're good to go.

**Comment by binoculars** [comment]
I added gulp-chmod on the google-cloud-function branch (see 8fc75c61567ef531a8714dc85eb8b2529b9e7536). Possibly fixes #13. I'm not sure if it's possibly to modify directory permissions in vinyl files, so it may have to happen during the zip process.

**Comment by binoculars** [comment]
No longer using async module.
